### PR TITLE
[TS migration] Migrate 'WorkspacePageWithSections' page to TypeScript

### PR DESCRIPTION
**Pull request by jayeshmangwani**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25185
PROPOSAL: https://github.com/Expensify/App/issues/25185#issuecomment-1883126975

### Tests

Prerequisite:
Go to Settings > Workspaces > Go to any workspace

Test 1:
1. On the worksapce page Press Settings option
2. Verify that Workspace Settings page loads and works as before

Test 2:
1. On the worksapce page Press Cards option
2. Verify that Cards page loads and works as before

Test 3:
1. On the worksapce page Press Reimbursements option
2. Verify that Reimbursements page loads and works as before

Test 4:
1. On the worksapce page Go to Reimbursements > Rate
2. Verify that Rate and Unit page loads and works as before

Test 5:
1. On the worksapce page Press Bills option
2. Verify that Bills page loads and works as before

Test 6:
1. On the worksapce page Press Invoices option
2. Verify that Invoices page loads and works as before

Test 7:
1. On the worksapce page Press Travel option
2. Verify that Travel page loads and works as before
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests
### QA Steps

Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/35371050/fa892bcc-2b0a-4b8b-93f2-b0c8a5266ca7

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/35371050/f67b3f55-5cb0-46ff-bc91-bbd944eebaac

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/35371050/5d1f9578-ccf5-46dd-88bf-3a0edfdf62b8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/35371050/77da5a1f-f7fb-40b0-a2b2-bebf096ea5dd

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/35371050/293a5dbf-0a71-4745-aebd-55e3623a51f4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/35371050/f9c77c51-b7e0-4322-beb2-928fc17cf134

</details>

### CODE REVIEW

**Comment by jayeshmangwani** [comment]
maybe I messed up with checklist that's why wrong reviewer got assigned, I didn't added a space after the Fixed issues section's $ sign may caused this, I am tagging you manually here @parasharrajat

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/workspace/withPolicy.tsx [hunk 1]
```diff
@@ -131,4 +131,4 @@ export default function <TProps extends WithPolicyProps, TRef>(WrappedComponent:
 }
 
 export {policyPropTypes, policyDefaultProps};
-export type {WithPolicyOnyxProps, WithPolicyProps};
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/withPolicy.tsx]: Now we are not using the PolicyRoute, so we does not have to do anything for shared type

File: src/pages/workspace/invoices/WorkspaceInvoicesPage.tsx [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+import type {PolicyRoute} from '@pages/workspace/withPolicy';
 import WorkspacePageWithSections from '@pages/workspace/WorkspacePageWithSections';
 import CONST from '@src/CONST';
-import type SCREENS from '@src/SCREENS';
 import WorkspaceInvoicesNoVBAView from './WorkspaceInvoicesNoVBAView';
 import WorkspaceInvoicesVBAView from './WorkspaceInvoicesVBAView';
 
-type WorkspaceInvoicesPageProps = StackScreenProps<SettingsNavigatorParamList, typeof SCREENS.WORKSPACE.INVOICES>;
+type WorkspaceInvoicesPageProps = {
+    route: PolicyRoute;
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/invoices/WorkspaceInvoicesPage.tsx]: added below comment, please check it and suggest my anything we need to change in it.

```
Defined route object that contains the policyID param, WorkspacePageWithSections is a common component for Workspaces and expect the route prop that includes the policyID
```

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 3]
```diff
@@ -1,86 +1,71 @@
...
 
-    /** Content to be added as fixed footer */
-    footer: PropTypes.element,
+        /** The text to display in the header */
+        headerText: string;
 
-    /** The guides call task ID to associate with the workspace page being shown */
-    guidesCallTaskID: PropTypes.string,
+        /** The route object passed to this page from the navigator */
+        route: PolicyRoute;
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx]: Thanks for the suggestion, replaced PolicyRoute with `RouteProp<{params: {policyID: string}}>;`

And we are using `RouteProp<{params: {policyID: string}}>;` at more then one place, so I think we should create a common type for it, but i am not sure about it where we should create it, please suggest me if we need to create a common type.

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 4]
```diff
@@ -101,17 +86,17 @@ function WorkspacePageWithSections({
...
+    const achState = reimbursementAccount?.achData?.state ?? '';
+    const isUsingECard = user?.isUsingExpensifyCard ?? false;
+    const policyID = route.params.policyID;
+    const policyName = policy?.name;
     const hasVBA = achState === BankAccount.STATE.OPEN;
-    const isUsingECard = lodashGet(user, 'isUsingExpensifyCard', false);
-    const policyID = lodashGet(route, 'params.policyID');
-    const policyName = lodashGet(policy, 'name');
-    const content = children(hasVBA, policyID, isUsingECard);
+    const content = children && children(hasVBA, policyID, isUsingECard);
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx]: oh yes, updated now

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 5]
```diff
@@ -123,6 +108,14 @@ function WorkspacePageWithSections({
         fetchData(shouldSkipVBBACall);
     }, [shouldSkipVBBACall]);
 
+    const shouldShow = () => {
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx]: updated!

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 6]
```diff
@@ -101,17 +86,17 @@ function WorkspacePageWithSections({
     shouldSkipVBBACall,
     user,
     shouldShowLoading,
-}) {
+}: WorkspacePageWithSectionsProps) {
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx:89]: updated!

we have a same user prop on WorkspacePageWithSections page and TestToolMenu,On TestToolMenu component we're using default value defined on the page, so I have extracted `USER_DEFAULT` to CONST file and now used same default value at both places.
https://github.com/Expensify/App/blob/0ebc6e49eb4f7217692d4d0cd0ded7c59f490253/src/components/TestToolMenu.tsx#L30-L32

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 7]
```diff
@@ -1,86 +1,71 @@
...
 import Navigation from '@libs/Navigation/Navigation';
 import * as PolicyUtils from '@libs/PolicyUtils';
-import * as ReimbursementAccountProps from '@pages/ReimbursementAccount/reimbursementAccountPropTypes';
-import userPropTypes from '@pages/settings/userPropTypes';
 import * as BankAccounts from '@userActions/BankAccounts';
 import ONYXKEYS from '@src/ONYXKEYS';
 import ROUTES from '@src/ROUTES';
+import type {Route} from '@src/ROUTES';
+import type {Policy, ReimbursementAccount} from '@src/types/onyx';
+import type UserOnyx from '@src/types/onyx/User';
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx]: updated!

File: src/components/TestToolMenu.tsx [hunk 8]
```diff
@@ -27,9 +27,7 @@ type TestToolMenuProps = TestToolMenuOnyxProps & {
     network: OnyxEntry<NetworkOnyx>;
 };
 
-const USER_DEFAULT: UserOnyx = {shouldUseStagingServer: undefined, isSubscribedToNewsletter: false, validated: false, isFromPublicDomain: false, isUsingExpensifyCard: false};
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/components/TestToolMenu.tsx:30]: sure, we will revert this change

[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/components/TestToolMenu.tsx:30]: I have reverted the changes that we have added in this PR for USER_DEFAULT. For  `src/components/TestToolMenu.tsx` file, we are not touching the USER_DEFAULT added by some other PR, please let me know if we need to update the USER_DEFAULT for this page to null

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 9]
```diff
@@ -1,86 +1,71 @@
...
 import useNetwork from '@hooks/useNetwork';
 import useThemeStyles from '@hooks/useThemeStyles';
-import compose from '@libs/compose';
 import BankAccount from '@libs/models/BankAccount';
 import Navigation from '@libs/Navigation/Navigation';
 import * as PolicyUtils from '@libs/PolicyUtils';
-import * as ReimbursementAccountProps from '@pages/ReimbursementAccount/reimbursementAccountPropTypes';
-import userPropTypes from '@pages/settings/userPropTypes';
 import * as BankAccounts from '@userActions/BankAccounts';
+import CONST from '@src/CONST';
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/workspace/WorkspacePageWithSections.tsx:18]: sorry, I was forgot to remove unnecessary imports, I will remove this in next commit

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


:ribbon: :eyes: :ribbon:  C+ reviewed

File: src/pages/workspace/WorkspacePageWithSections.tsx [hunk 10]
```diff
@@ -123,6 +107,14 @@ function WorkspacePageWithSections({
         fetchData(shouldSkipVBBACall);
     }, [shouldSkipVBBACall]);
 
+    const shouldShow = useMemo(() => {
+        if (isEmptyObject(policy)) {
+            return true;
+        }
+
+        return !PolicyUtils.isPolicyAdmin(policy) || PolicyUtils.isPendingDeletePolicy(policy);
+    }, [policy]);
```
Comment (aldo-expensify) [inline src/pages/workspace/WorkspacePageWithSections.tsx:116]: NAB: IMO the code here don't justify a `useMemo` as it is very lean. Just a couple of `!==` checks and `Object.keys(obj ?? {}).length === 0`. I think `useMemo` makes the code less readable and maybe even slower here :P

**Review by aldo-expensify — APPROVED** [review]

**Pull request merged**